### PR TITLE
Implement map-reduction of BQSR (fix #184)

### LINKED ISSUE #184: Implement parallelization for BQSR
Cf. http://gatkforums.broadinstitute.org/wdl/discussion/1988/a-primer-on-parallelism-with-the-gatk

and http://gatkforums.broadinstitute.org/gatk/discussion/1919/parallelizing-base-quality-score-recalibration for specifics wrt BQSR

### PR DESCRIPTION
**Pull request by smondet**
Still testing.

PS: sorry for the whitespacy diff 🙏

---
This change is [Reviewable](https://reviewable.io/reviews/hammerlab/biokepi/286)

### CODE REVIEW

File: src/bfx_tools/gatk.ml [hunk 1]
```diff
@@ -497,6 +505,25 @@ let base_quality_score_recalibrator
...
+    Machine.get_reference_genome run_with input_bam#product#reference_build in
+  let run_on_region region =
+    base_quality_score_recalibrator_on_region ~configuration
+      ~run_with ~input_bam ~result_prefix ~region
+  in
+  match how with
+  | `Region region -> run_on_region region
+  | `Map_reduce ->
+    let all_nodes =
+      List.map (Reference_genome.major_contigs reference)
```
Comment (ihodes) [inline src/bfx_tools/gatk.ml]: Does this mean it'll not run on other regions? Seems important that we don't leave those behind (I'm guessing that we don't run other parallelized tools over non-major-contigs?)

Comment (smondet) [inline src/bfx_tools/gatk.ml]: Yes all region-based map-reducing uses the `major_contigs` function (then the user can put whatever they want behind that; it's just a property of the reference genome).

**Comment by arahuja** [comment]
Doesn't this throw off  the calibration? Is there enough data within a chromosome for all of the covariates? And this will give different results compared to running all together?

Perhaps the first `BaseRecalibrator` step should be run on all the data and then `PrintReads` should be run in parallel?

**Comment by ihodes** [comment]
@arahuja Broad seems to think it's okay: https://github.com/hammerlab/biokepi/issues/184

File: src/bfx_tools/gatk.ml (hunk 1 above)
Comment (ihodes) [inline src/bfx_tools/gatk.ml]: Is there an easy way for the user to specify that it should be run over all the contigs? Or would they need to enumerate the name of all the other small contigs/unmapped regions in the reference in order to get this behavior? (And thereby creating a "new" `Reference_genome.t`?)

**Comment by arahuja** [comment]
@ihodes commented at https://github.com/hammerlab/biokepi/issues/184#issuecomment-224674310

**Comment by smondet** [comment]
@ihodes @arahuja Also, I don't know if related but with the PR my strelka fails with: https://sites.google.com/site/strelkasomaticvariantcaller/home/faq#TOC-The-workflow-reported-an-error-described-as-ERROR:-multiple-suggested-alignments-for-read-are-not-same-strand-

**Comment by smondet** [comment]
@ihodes @arahuja so we confirm that this //ization method (= per contig) goes nowhere?

**Comment by ihodes** [comment]
Yeah unfortunately I think we should avoid merging this, and punt until we can work out how to properly parallelize it with their tools.

**Comment by smondet** [comment]
OK 😿

**Comment by ihodes** [comment]
😢

**Pull request closed without merging**